pkg/reconciler/knativeserving/common/images.go

```diff
@@ -20,60 +20,62 @@ import (
 
 	mf "github.com/jcrossley3/manifestival"
 	"go.uber.org/zap"
-	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/client-go/kubernetes/scheme"
 	caching "knative.dev/caching/pkg/apis/caching/v1alpha1"
+	"knative.dev/pkg/apis/duck"
+	"knative.dev/pkg/apis/duck/v1"
 	servingv1alpha1 "knative.dev/serving-operator/pkg/apis/serving/v1alpha1"
 )
 
 func init() {
 	caching.AddToScheme(scheme.Scheme)
+	v1.AddToScheme(scheme.Scheme)
 }
 
 var (
 	// The string to be replaced by the container name
 	containerNameVariable = "${NAME}"
 )
 
-func DeploymentTransform(instance *servingv1alpha1.KnativeServing, log *zap.SugaredLogger) mf.Transformer {
-	return func(u *unstructured.Unstructured) error {
-		// Update the deployment with the new registry and tag
-		if u.GetKind() == "Deployment" {
-			return updateDeployment(instance, u, log)
-		}
-		return nil
-	}
-}
-
+// ImageTransform updates image with a new registry and tag
 func ImageTransform(instance *servingv1alpha1.KnativeServing, log *zap.SugaredLogger) mf.Transformer {
 	return func(u *unstructured.Unstructured) error {
-		// Update the image with the new registry and tag
-		if u.GetAPIVersion() == "caching.internal.knative.dev/v1alpha1" && u.GetKind() == "Image" {
-			return updateCachingImage(instance, u)
+		switch u.GetKind() {
+		case "Deployment", "DaemonSet":
+			return updateDeployment(instance, u, log)
+		case "Image":
+			if u.GetAPIVersion() == "caching.internal.knative.dev/v1alpha1" {
+				return updateCachingImage(instance, u)
+			}
+			return nil
+		default:
+			return nil
 		}
-		return nil
 	}
 }
 
 func updateDeployment(instance *servingv1alpha1.KnativeServing, u *unstructured.Unstructured, log *zap.SugaredLogger) error {
-	var deployment = &appsv1.Deployment{}
-	err := scheme.Scheme.Convert(u, deployment, nil)
-	if err != nil {
-		log.Error(err, "Error converting Unstructured to Deployment", "unstructured", u, "deployment", deployment)
+	specData := &v1.WithPod{}
+	if err := duck.FromUnstructured(u, specData); err != nil {
+		log.Error(err, "error converting Unstructured to Duck type")
 		return err
 	}
-
 	registry := instance.Spec.Registry
 	log.Debugw("Updating Deployment", "name", u.GetName(), "registry", registry)
 
-	updateDeploymentImage(deployment, &registry, log)
-	deployment.Spec.Template.Spec.ImagePullSecrets = addImagePullSecrets(
-		deployment.Spec.Template.Spec.ImagePullSecrets, &registry, log)
-	err = scheme.Scheme.Convert(deployment, u, nil)
+	updateDeploymentImage(specData.Spec.Template.Spec.Containers, specData.GetName(), &registry, log)
+	specData.Spec.Template.Spec.ImagePullSecrets = addImagePullSecrets(
+		specData.Spec.Template.Spec.ImagePullSecrets, &registry, log)
+
+	unstructuredData, err := duck.ToUnstructured(specData.DeepCopyObject().(duck.OneOfOurs))
```

Comment:
Oh I see, you need the `selector` in PodSpecable to successfully serialize back here? 🤔 

Can you instead use the existing unstructured and only override things? I.e. take the transformed podSpec and set it into the unstructured element at `Spec.Template` or something like that?

Reply:
Do you mean this https://github.com/knative/serving-operator/pull/281/files#diff-411a842e3f404ed5adc1b082ea595d72R78 directly instead of duck.ToUnstructured ?

Reply:
No I don't think that'd work either. Lemme try if what I though actually works before I bother you more with it 😂

Reply:
Yep that also won't work

Reply:
In pseudo-code, we need something like this:

```
podSpecable := duck.FromUnstructured(u)

updateImages(podSpecable)
updatePullSecret(podSpecable)

unstructured.SetNestedField(u.Object, podSpecable.Spec.Template, "spec", "template")
```

Or alternatively

```
podSpecable := duck.FromUnstructured(u)
afterChanges := podSpecable.DeepCopy()

updateImages(afterChanges)
updatePullSecret(afterChanges)

patch := generatePatch(podSpecable, afterChanges)
applyPatch(u, patch)
```

That way we keep the fields of the unstructured intact that PodSpecable doesn't capture (see selector etc.) while also having the profit of using just one type. I haven't gotten a working version of the above together yet though.